src/drivers/intel/cavs/hda-dma.c

```diff
@@ -383,9 +383,11 @@ static int hda_dma_link_copy_ch(struct dma *dma, struct hda_chan_data *chan,
 		hda_update_bits(dma, chan->index,
 				DGCS, DGCS_BOR, DGCS_BOR);
 
-	/* make sure that previous transfer is complete */
-	while (hda_dma_get_free_size(dma, chan->index) < bytes)
-		idelay(PLATFORM_DEFAULT_DELAY);
+	/* make sure that previous transfer is complete (playback only) */
+	if (chan->direction == DMA_DIR_MEM_TO_DEV) {
+		while (hda_dma_get_free_size(dma, chan->index) < bytes)
```

Comment:
What happens if free size is always < bytes  ? i.e. due to some HW issue or some race with XRUN or stream STOP cmd ? We probably want a timeout in these loops (maybe double the length of time it would take bytes to be free) and then exit with an error.

Reply:
I already have that change prepared. I will create a separate PR once I figure out the timing for the timeout as it requires testing in various cases.